src/github.com/couchbase/sync_gateway/db/change_listener.go

```diff
@@ -174,9 +195,19 @@ func (listener *changeListener) NewWaiterWithChannels(chans base.Set, user auth.
 }
 
 // Waits for the changeListener's counter to change from the last time Wait() was called.
-func (waiter *changeWaiter) Wait() bool {
-	waiter.lastCounter = waiter.listener.Wait(waiter.keys, waiter.lastCounter)
-	return waiter.lastCounter > 0
+func (waiter *changeWaiter) Wait() uint32 {
+	var terminateCheck bool
+	waiter.lastCounter, terminateCheck = waiter.listener.Wait(waiter.keys, waiter.lastCounter)
+	countChanged := waiter.lastCounter > 0
+
+	if terminateCheck {
+		return 2
+	} else if countChanged {
```

Comment:
This needs to be the first one in the if statement, I think - if both countChanged and terminateCheck are true, we need to treat as a regular notify.